### DISCUSSION TITLE
Opinions on PEP 724?

### DISCUSSION THREAD

**Discussion by hauntsaninja** [sqlalchemy/sqlalchemy#10963] [Ideas]
(apologies for the misuse of this issue tracker)

I'm curious if @zzzeek or other maintainers here have an opinion on https://peps.python.org/pep-0724/ , since SQLAlchemy appears to be one of the heaviest open source users of TypeGuard. Please feel free to chime in at https://discuss.python.org/t/pep-724-stricter-type-guards/ :-)

**Comment by CaselIT** [comment]
Hi,

I've skimmed that pep, and while I can't think at specific cases where it would be useful on sqlalchemy, I honestly though that `TypeGuard` already worked like that. (I haven't yet used it a lot, so I did not remember the specifics of the current behaviour).
So my personal opinion would be in favour of that PEP. Let's also wait for @zzzeek opinion that has implemented most of the typing support

A suggestion on the pep description: It mat make sense to show an example of what happen on an union of 3 or more argument, something like
```py
def is_str(arg: Any)->TypeGuard[str]: ... # does not make much sense since it's isinstance(value,str)
def fn(value: str|int|date|datetime):
  if is_str(value):
    ... # value is str
  else:
    ... # value here is int|date|datetime (I'm assuming it, would be a useful example in the pep)
```

moving to discussion since it's not an actionable issue on sqlalchemy part

**Comment by zzzeek** [comment]
this seems like a fine enhancement to typing tools, I dont have much opinion on it, we use TypeGuard only (?) internally and it does what we need for now, if TypeGuard becomes "better" for the false cases, seems good.

the PEP does not look controversial and the proposers are basically "the typing guys", looks like it will eventually show up in tooling?
